incorporating-accessibility-into-your-dev-process/focus-tests/solution/src/App.test.tsx

```diff
@@ -5,37 +5,122 @@ Exercise: finish the “gives focus to the first invalid input” test below.
+    await user.type(screen.getByLabelText(/first name:/i), "John");
+    await user.click(
+      screen.getByRole("button", { name: /validate full name/i }),
+    );
+
+    // Last name input should be focused now since it’s still invalid
+    expect(screen.getByLabelText(/last name:/i)).toHaveFocus();
+  });
+
+  it("displays an error when submitting the form with empty inputs", async () => {
+    render(<App />);
+
+    await userEvent.click(
+      screen.getByRole("button", { name: /validate full name/i }),
+    );
 
-    it('gives focus to the first invalid input', async () => {
-        const user = userEvent.setup();
+    expect(
+      screen.getByText(/the “first name” field is required./i),
+    ).toBeInTheDocument();
+    expect(
+      screen.getByText(/the “last name” field is required./i),
+    ).toBeInTheDocument();
+  });
 
-        render(<App />);
+  it("focuses on the first name input when it is invalid and submitted", async () => {
+    render(<App />);
 
-        // Click the button without filling out any fields
-        await user.click(screen.getByRole('button', { name: /validate full name/i }));
+    const firstNameInput = screen.getByLabelText(/first name:/i);
+    firstNameInput.focus();
+    await userEvent.tab(); // This should move focus away from the first name input
+    expect(firstNameInput).not.toHaveFocus();
 
-        // First name input should be focused since it’s invalid
-        expect(screen.getByLabelText(/first name:/i)).toHaveFocus();
+    await userEvent.click(
+      screen.getByRole("button", { name: /validate full name/i }),
+    );
+    expect(firstNameInput).toHaveFocus();
+  });
 
-        // Fill out the first name only
-        await user.type(screen.getByLabelText(/first name:/i), 'John');
-        await user.click(screen.getByRole('button', { name: /validate full name/i }));
+  it("focuses on the last name input when it is invalid and submitted", async () => {
+    render(<App />);
 
-        // Last name input should be focused now since it’s still invalid
-        expect(screen.getByLabelText(/last name:/i)).toHaveFocus();
+    await userEvent.type(screen.getByLabelText(/first name:/i), "John");
+
+    const lastNameInput = screen.getByLabelText(/last name:/i);
+    lastNameInput.focus();
+    await userEvent.tab(); // This should move focus away from the last name input
+    expect(lastNameInput).not.toHaveFocus();
+
+    await userEvent.click(
+      screen.getByRole("button", { name: /validate full name/i }),
+    );
+    expect(lastNameInput).toHaveFocus();
+  });
+
+  it("displays a success message with the full name when the form is submitted with valid inputs", async () => {
+    render(<App />);
+
+    await userEvent.type(screen.getByLabelText(/first name:/i), "John");
+    await userEvent.type(screen.getByLabelText(/last name:/i), "Doe");
+    await userEvent.click(
+      screen.getByRole("button", { name: /validate full name/i }),
+    );
+
+    await waitFor(() => {
```

Comment:
the findBy api would simplify this a bit, its basically a wrapper around `waitFor` with a `getBy...`

IIRC its recommended to use it instead of the waitFor getBy... but i can't find the document that says that

https://testing-library.com/docs/dom-testing-library/api-async/#findby-queries